### PR TITLE
Node doctor

### PR DESCRIPTION
**Pull request by jakecoll**
Add some nodes by name and it will diagnose based known failure causes in the chameleon admin docs.

I also refactored the osrest code and added the Placement API.

### CODE REVIEW

File: hammers/scripts/node_doctor.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+    node-doctor <node_name>
+
+'''
+import argparse
+import re
+import sys
+
+from hammers.scripts.ironic_error_resetter import cureable_nodes
+from hammers.scripts.undead_instances import find_unbound_instances
```
Comment (diurnalist) [inline hammers/scripts/node_doctor.py]: If I came back to this codebase, I would not expect scripts to use functions defined within eachother. I think we should leave a comment on the `find_unbound_instances` function saying something like "NOTE: if you modify/delete this function, be sure to check its usage in node_doctor". Or, simply copy the implementation here and remove the coupling.

File: hammers/osrest/base.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        if self.extra_headers:
+            headers.update(self.extra_headers)
+        if content_type:
+            headers.update({'Content-Type': content_type})
+
+        return headers
+
+    def get(self, auth, path, params=None):
+        response = requests.get(url=auth.endpoint(self.service) + path,
```
[the remark this answers is not in the document]
Comment (jakecoll) [reply hammers/osrest/base.py]: We use the keystoneauth session and adapter in the openstack exporter. It's a bit finnicky depending on the version. May be worth incorporating at some point.

[the remark this answers is not in the document]
Comment (jakecoll) [reply hammers/osrest/base.py]: auth is a custom class in hammers. It's not a keystoneauth object.

File: hammers/osrest/blazar.py [hunk 3]
```diff
@@ -1,80 +1,55 @@
...
 def host(auth, host_id):
     """Retrieves host by ID."""
-    response = requests.get(
-        url=auth.endpoint('reservation') + '/os-hosts/{}'.format(host_id),
-        headers={'X-Auth-Token': auth.token}
-    )
-    response.raise_for_status()
-    bhost = response.json()['host']
-    return bhost
+    response = API.get(auth, '/os-hosts/{}'.format(host_id))
```
[the remark this answers is not in the document]
Comment (jakecoll) [reply hammers/osrest/blazar.py]: lol

File: hammers/scripts/node_doctor.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+
+def available_nodes(nodes):
+    return [
+        nid for nid, node in nodes.items()
+        if not node['maintenance'] and node['provision_state'] == "available"]
+
+
+def node_in_error_state(nodes):
+    error_state_nodes = cureable_nodes(nodes)
```
[the remark this answers is not in the document]
Comment (jakecoll) [reply hammers/scripts/node_doctor.py]: I'll change to just check for error state.

File: hammers/scripts/node_doctor.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+    for node in error_state_nodes:
+        nodes[node]['ailements'].append("error_state")
```
[the remark this answers is not in the document]
Comment (jakecoll) [reply hammers/scripts/node_doctor.py]: fixed

File: hammers/scripts/node_doctor.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+
+def node_maintenance_state_error(auth, nodes):
+    pattern = re.compile(MAINTENANCE_LEASE_REGEX)
```
[the remark this answers is not in the document]
Comment (jakecoll) [reply hammers/scripts/node_doctor.py]: The idea is to target nodes in maintenance mode without a corresponding maintenance lease. This is the "bad" state we need to address whereas maintenance mode + maintenance lease is valid.

File: hammers/scripts/node_doctor.py (hunk 1 above)
Comment (jakecoll) [reply hammers/scripts/node_doctor.py]: I'll add functionality directly into node doctor

**Pull request merged**